korim/src/commonMain/kotlin/com/soywiz/korim/bitmap/BitmapSlice.kt

```diff
@@ -281,68 +329,313 @@ abstract class BmpSlice(
+                brX.clamp(pMinX, pMaxX), brY.clamp(pMinY, pMaxY),
+                blX.clamp(pMinX, pMaxX), blY.clamp(pMinY, pMaxY)
+            )
+            val minX = coords.slice(0..6 step 2).minOrNull()!!
+            val minY = coords.slice(1..7 step 2).minOrNull()!!
+
+            for (i in 0..6 step 2) {
+                if (coords[i] == minX && coords[i + 1] == minY) {
+                    val rotated = imageOrientation.rotation == ImageOrientation.Rotation.R90 || imageOrientation.rotation == ImageOrientation.Rotation.R270
+                    if ((tlX == trX && !rotated) || (tlX != trX && rotated)) {
+                        // rotated
+                        bounds.height = ((coords[(i + 4) % 8] - coords[i]) * bmpBase.width).roundToInt()
+                        bounds.width = ((coords[(i + 5) % 8] - coords[i + 1]) * bmpBase.height).roundToInt()
+                    } else {
+                        bounds.width = ((coords[(i + 4) % 8] - coords[i]) * bmpBase.width).roundToInt()
+                        bounds.height = ((coords[(i + 5) % 8] - coords[i + 1]) * bmpBase.height).roundToInt()
+                    }
+                    break
+                }
+            }
+
+            this.bmpCoords = BmpCoordsWithInstance<ISizeInt>(bounds.size, coords[0], coords[1], coords[2], coords[3], coords[4], coords[5], coords[6], coords[7])
+                .withImageOrientation(imageOrientation)
+        }
+    }
+
     val trimmed: Boolean = virtFrame != null
-    override val frameOffsetX: Int = virtFrame?.x ?: 0
-    override val frameOffsetY: Int = virtFrame?.y ?: 0
     override val frameWidth: Int = virtFrame?.width ?: bounds.width
     override val frameHeight: Int = virtFrame?.height ?: bounds.height
 
-	var parent: Any? = null
+    var parent: Any? = null
 
-    override val tl_x = p0.x.toFloat()
-    override val tl_y = p0.y.toFloat()
-    override val tr_x = p1.x.toFloat()
-    override val tr_y = p1.y.toFloat()
-    override val br_x = p2.x.toFloat()
-    override val br_y = p2.y.toFloat()
-    override val bl_x = p3.x.toFloat()
-    override val bl_y = p3.y.toFloat()
+    override val tl_x: Float get() = this.bmpCoords.tl_x
+    override val tl_y: Float get() = this.bmpCoords.tl_y
+    override val tr_x: Float get() = this.bmpCoords.tr_x
+    override val tr_y: Float get() = this.bmpCoords.tr_y
+    override val br_x: Float get() = this.bmpCoords.br_x
+    override val br_y: Float get() = this.bmpCoords.br_y
+    override val bl_x: Float get() = this.bmpCoords.bl_x
+    override val bl_y: Float get() = this.bmpCoords.bl_y
 
+    @Deprecated("Use imageOrientation")
+    val rotated: Boolean = imageOrientation == ImageOrientation.ROTATE_90 || imageOrientation == ImageOrientation.ROTATE_270
     val rotatedAngle: Int = 0
 
+    private val pixelOffsets: IntArray by lazy {
+        val x = (tl_x * baseWidth).roundToInt()
+        val y = (tl_y * baseHeight).roundToInt()
+        if (tl_x == tr_x) {
```

Comment:
I have seen this comparison in a couple of places. What's the meaning of that comparison? When that is true how is the BmpSlice?

Can we have something like:

```kotlin
class BmpSlice {
  val isSomething: Boolean get() = tl_x == tr_x

  ...
  if (isSomething) {
    ...
  }
}
```

Reply:
That is, if the `BmpSlice` is rotated 90° CW or CCW relative to the Bitmap. Then `tl_x` is equal to `tr_x` and `width` and `height` must be flipped in calculations between `bounds` and `bmpCoords`. (That is not `true` for example for a slice of a slice where both are rotatet 90°)

I suggest this name and the comparsion should be also changed to be correct.
``` kotlin
val isRotatedInBmpDeg90 = (width > 1 && tl_x == tr_x) || (height > 1 && tl_y == bl_y)
```